File: app/src/main/java/com/chess/cryptobot/task/SerialExecutor.java
package com.chess.cryptobot.task;

import java.util.ArrayDeque;
import java.util.concurrent.Executor;

import static android.os.AsyncTask.THREAD_POOL_EXECUTOR;

public class SerialExecutor implements Executor {
    private final ArrayDeque<Runnable> mTasks = new ArrayDeque<>();
    private Runnable mActive;

    public synchronized void execute(final Runnable r) {
        mTasks.offer(() -> {
            try {
                r.run();
            } finally {
                scheduleNext();
            }
        });
        if (mActive == null) {
            scheduleNext();
        }
    }

    private synchronized void scheduleNext() {
        if ((mActive = mTasks.poll()) != null) {
            THREAD_POOL_EXECUTOR.execute(mActive);
        }
    }
}

File: app/src/main/java/com/chess/cryptobot/task/SerialExecutor.kt
package com.chess.cryptobot.task

import android.os.AsyncTask
import java.util.*
import java.util.concurrent.Executor

class SerialExecutor : Executor {
    private val mTasks = ArrayDeque<Runnable>()
    private var mActive: Runnable? = null
    @Synchronized
    override fun execute(r: Runnable) {
        mTasks.offer(Runnable {
            try {
                r.run()
            } finally {
                scheduleNext()
            }
        })
        if (mActive == null) {
            scheduleNext()
        }
    }

    @Synchronized
    private fun scheduleNext() {
        if (mTasks.poll().also { mActive = it } != null) {
            AsyncTask.THREAD_POOL_EXECUTOR.execute(mActive)
        }
    }
}